Add basic docs with sphinx

niksirbi opened the pull request:
This will tackle issues #25 and #43

samcunliffe commented on .github/workflows/publish_docs.yml:
```diff
@@ -0,0 +1,18 @@
+      - '*'
+
+jobs:
+  publish_sphinx_docs:
```
Should this be merged with `check_doc.yml` and linked with a `needs`? I.e....
```
check_sphinx_docs → publish_sphinx_docs
```

niksirbi replied:
I ended up getting rid of `check_docs.yml` for now (see reasons in my extensive comment)

samcunliffe commented:
That was quick!

niksirbi commented:
## Summary of changes
* Added necessary Sphinx files in `docs` folder. The important ones are listed below:
    * `docs/requirements.txt`: documentation build requirements
    * `docs/source/conf.py`: configures the sphinx build (including theme, extensions, URL, etc.)
    * `docs/source/index.rst`: this corresponds to the docs homepage, and must contain a table-of-contents with refs to other files in the `docs/source` directory. Additional docs paged can be added as separate `.md` or `rst` files.
* Added a github workflow `.github/workflows/publish_docs` that is triggered by pushes to `main`. This does the following:
    * installs Sphinx and the other docs build requirements specified in `docs/requirements.txt`
    * build the contents of `docs/source` into `HTML` and transfers the build to a `gh-pages` branch.
    * The rendered build is available at: https://sainsburywellcomecentre.github.io/WAZP/ (the reason you can already see it up, is because I temporarily enabled the above workflow for this branch as well, will disable before merge)
* Transferred most of the previous contents of the `README.md` to a `CONTRIBUTING.md` and significantly expanded the "how to contribute" information. 
* Expanded the `README.md` file by adding badges, and instructions for installing and launching WAZP.
* I made the contents of the `README.md` appear on the homepage of the [docs website](https://sainsburywellcomecentre.github.io/WAZP/), while the `CONTRIBUTING.md` file is rendered as a separate page on the same website. I accomplished this via linking to the corresponding repo files from inside `docs/source` (to avoid content duplication).
    
 ## Remaining issues
 * Originally, I wanted to add [an additional workflow](https://github.com/neuroinformatics-unit/actions/tree/main/check_sphinx_docs) that would be triggered upon pull requests and would check whether the PR breaks the docs Sphinx build. However, that particular workflow gave me many problems (because the action I was relying on no longer seems to be actively maintained, see [this issue](https://github.com/neuroinformatics-unit/actions/issues/8)) and I have abandoned it for now. For the future, we will still need an action that checks the docs build upon PRs (any ideas @samcunliffe ?)
 * Currently I have not included any API docs (I don't we need that for now, since the users be interacting with the dashboard, not directly calling Python functions). That said, it will be easy to auto-generate API docs with Sphinx, should we need them in the future.
 * Actual instructions on how to use the dashboard are currently missing, but I figured that these should come in separate PRs. These could take the form of adding more`.md` files to `docs/source`, and including them in the table-of-contents in `docs/source/index.rst`

samcunliffe commented:
> The rendered build is available at: https://sainsburywellcomecentre.github.io/WAZP/ 

😻 It looks really nice 👍 👍

samcunliffe commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,148 @@
+- Please submit _draft_ pull requests as early as possible to allow for discussion.
+- One approval of a PR (by a repo owner) is enough for it to be merged.
+- Unless someone approves the PR with optional comments, the PR is immediately merged by the approving reviewer.
+- We typically merge via "Squash and Merge" on GitHub to maintain a clean commit history.
```
```suggestion
```
To my taste we should just make this the rule.

I've just enforced this in settings (and therefore we don't need a sentence ☝️). In case you want to revert, it's the general settings tab at the bottom just before the danger zone under "Pull requests".

samcunliffe replied:
![Screenshot 2023-03-06 at 10 40 10](https://user-images.githubusercontent.com/1836192/223087490-84f8a3b6-334a-4dd4-a4ab-532acaa1fe93.png)

samcunliffe commented on docs/source/readme_link.md:
````diff
@@ -0,0 +1,2 @@
+```{include} ../../README.md
+```
````
Is it needed? I would expect you can just directly... 
```
.. include:: ../../README.md
   :parser: myst_parser.sphinx_
```
... in index.rst. (Also applies to CONTRIBUTING.)

niksirbi replied:
Thanks for that.
It worked for README (so I removed the unnecessary extra file), but not for CONTRIBUTING. I guess because the latter is specified in the TOC (not simply included in an existing file). I could be missing sth, but having an extra "link" file in `docs/source` is the only solution that works for me so far

samcunliffe replied:
Ah, I see the problem. [I've just checked out `add-docs` and had a play]
Your way is the most explicit that I've come up with, and that works. And therefore the best.

samcunliffe commented on pyproject.toml:
```diff
@@ -73,7 +73,7 @@ include-package-data = true
 [tool.setuptools.packages.find]
 include = ["wazp*"]
-exclude = ["tests*"]
+exclude = ["tests*", "docs*"]
```
👍

samcunliffe commented on MANIFEST.in:
```diff
@@ -1,11 +1,12 @@
 
 recursive-exclude sample_project *.avi
 recursive-exclude sample_project *.h5
+recursive-exclude docs *
```
👍

samcunliffe commented on docs/source/conf.py:
```diff
@@ -0,0 +1,121 @@
+sys.path.insert(0, os.path.abspath("../.."))
+
+project = "wazp"
+copyright = "2022, UCL"
+author = "UCL"
```
```suggestion
copyright = "2022, The Sainsbury Wellcome Centre"
author = "The Sainsbury Wellcome Centre and The Centre for Advanced Research Computing, UCL"
```
[The license is the SWC](https://github.com/SainsburyWellcomeCentre/WAZP/blob/main/LICENSE#L3). Not that it matters if SWC is a member of UCL but I'd keep it consistent.

And maybe "Author" should be both? What do you think?

niksirbi replied:
I'm fine with this, but @adamltyson may know more

adamltyson replied:
I don't think what we write here has any effect upon IP law, but AFAIK UCL (as the legal entity) owns copyright, so everything should probably reflect that.

adamltyson replied:
I think copyright - UCL, and author - SWC & ARC is nice.

niksirbi replied:
Should I also change the copyright to UCL in BSD-3?

adamltyson replied:
Yup

niksirbi replied:
Done

samcunliffe commented on README.md:
```diff
@@ -1,45 +1,57 @@
+⚖️ [BSD 3-Clause](https://opensource.org/licenses/BSD-3-Clause)
 
-### Template
+## Template
 This package layout and configuration (including pre-commit hooks and GitHub actions) have been copied from the [python-cookiecutter](https://github.com/SainsburyWellcomeCentre/python-cookiecutter) template.
```
(Soft suggestion) Perhaps this is better in CONTRIBUTING?

niksirbi replied:
You are right, I moved it now

samcunliffe commented on README.md:
```diff
@@ -1,45 +1,57 @@
+[![docs](https://img.shields.io/website?down_color=red&down_message=down&label=docs&up_color=brightgreen&up_message=up&url=https%3A%2F%2Fsainsburywellcomecentre.github.io%2FWAZP%2F)](https://sainsburywellcomecentre.github.io/WAZP/)
+[![Imports: isort](https://img.shields.io/badge/%20imports-isort-%231674b1?style=flat&labelColor=ef8336)](https://pycqa.github.io/isort/)
+[![Code style: black](https://img.shields.io/badge/code%20style-black-000000.svg)](https://github.com/python/black)
+[![pre-commit](https://img.shields.io/badge/pre--commit-enabled-brightgreen?logo=pre-commit&logoColor=white)](https://github.com/pre-commit/pre-commit)
```
Badges are cool.

samcunliffe commented on README.md:
```diff
@@ -1,45 +1,57 @@
-## How to contribute
-### Setup
-* Clone the repository and install the package in editable mode (including all `dev` dependencies):
+WAZP is a dashboard built with [Dash-Plotly](https://dash.plotly.com/) for analysing animal tracking data. It is designed to work with pose estimation output from [DeepLabCut](https://github.com/DeepLabCut/DeepLabCut).
```
```suggestion
WAZP is a dashboard built with [Dash-Plotly](https://dash.plotly.com/) for analysing animal tracking data. It can display pose estimation output from [DeepLabCut](https://github.com/DeepLabCut/DeepLabCut).
```
A bit too much qualification for my taste. And let's stick to what *is*. We will implement fancier things and support other pose estimation tools. Then we update the docs.

samcunliffe commented on README.md:
```diff
@@ -1,45 +1,57 @@
+
+## Installation
+
+It is recommended to install WAZP inside a [conda](https://docs.conda.io/en/latest/) environment.
```
```suggestion
We recommend you install WAZP inside a [conda](https://docs.conda.io/en/latest/) environment.
```

samcunliffe commented on README.md:
```diff
@@ -1,45 +1,57 @@
+## Installation
+
+It is recommended to install WAZP inside a [conda](https://docs.conda.io/en/latest/) environment.
+Once you have `conda` installed, the following commands will create and activate a suitable environment, named `wazp-env`.
```
```suggestion
Once you have `conda` installed, the following will create and activate an environment. You can call your environment whatever you like, we've used `wazp-env`.
```

samcunliffe commented on README.md:
````diff
@@ -1,45 +1,57 @@
+```
+
+Next clone the GitHub repository, navigate to the local repository folder, and install WAZP.
+
+```sh
 git clone https://github.com/SainsburyWellcomeCentre/WAZP
 cd WAZP
-pip install -e '.[dev]'
+pip install .
 ```
````
I can't leave a fancy suggestion here because my idea straddles a deleted line. 
Can't you use the fancy `pip` install direct from GitHub? A one-liner.

niksirbi replied:
Of course putting it on PyPI is the plan. But I was thinking we were delaying this until we have a minimally functional end-to-end product? 
I'm personally fine with releasing it as is though. In that case, however, what should be the version (currently it's `v0.1` + some dev tags and commit hashes that `setuptools_scm` adds.

adamltyson replied:
I generally favour releasing on PyPI at this stage, so that the users don't get one set of installation instructions that then get replaced by others. Version number doesn't really matter as long as it's <1 and we stick to semver.

niksirbi replied:
I will try to make the first PyPI release, right after this PR is merged

samcunliffe commented on README.md:
```diff
@@ -1,45 +1,57 @@
-* Create a new branch for your changes, make your changes and commit them.
+## Launching the dashboard
+
+You can launch the dashboard by opening a terminal and running the following command from the root of the repository:
```
```suggestion
Once installed, launch the dashboard by running the following command from the root of the repository:
```
(They're presumably still in a terminal)

samcunliffe commented on README.md:
```diff
@@ -1,45 +1,57 @@
 
-### How to run tests and type checks locally
-For debugging purposes, you may also want to run the tests and the type checks locally, before pushing. This can be done with the following commands:
+Alternatively, on Linux and MacOS, you can also run the following command:
```
```suggestion
If you're on Linux or MacOS, you can instead run:
```

samcunliffe reviewed (approved):
This looks really good 😻!!!

I've left some pedantic comments for our new user-facing README.

I didn't comb through CONTRIBUTING quite so thoroughly because I expect it will evolve a fair bit and this is already an excellent base from which to build.

> Currently I have not included any API docs (I don't we need that for now, since the users be interacting with the dashboard, not directly calling Python functions). That said, it will be easy to auto-generate API docs with Sphinx, should we need them in the future.

I'd put them in right now if it's easy. Can you simply `autodoc` the lot? If it's a pain then merge this PR first.

> Actual instructions on how to use the dashboard are currently missing, but I figured that these should come in separate PRs. These could take the form of adding more.md files to docs/source, and including them in the table-of-contents in docs/source/index.rst

Agreed.

niksirbi commented:
> When you mention GH issues on GH PRs and/or comments, you can use [some magic words](https://github.blog/2013-01-22-closing-issues-via-commit-messages/) and the issue gets linked/closed.

I actually knew this, but deliberately didn't use one of the magic words (since when I opened the PR I wasn't sure which issues would be closed by it 😈 ).

Now however I can say with confidence that this PR closes issue #25 😉

niksirbi commented on .github/workflows/publish_docs.yml:
```diff
@@ -0,0 +1,19 @@
+    steps:
+    - uses: neuroinformatics-unit/actions/publish_sphinx_docs@v1
+      with:
+        secret_input: ${{ secrets.GITHUB_TOKEN }}
```
That's a neat solution. I will try to implement the two-job workflow upstream, in our actions-repo, since this issue affects all our other repos-with-docs as well.

niksirbi replied:
I implemented the new two job workflow upstream [in this PR](https://github.com/neuroinformatics-unit/actions/pull/13).
To test things, I made both the build and the deploy jobs run from this branch, and it seems to have worked.

Before merging the current PR, I will configure it so that all merges to main and all pull requests trigger the build action, but only pushing/merging to main triggers a deployment to gh-pages. In the future, we may change this to only be triggered by pushing release tags. Does that sound reasonable @samcunliffe ?

samcunliffe replied:
All fine for me. I leave it for you to merge when ready.
